theano/gof/opt.py

```diff
@@ -533,19 +544,64 @@ def process_node(self, fgraph, node):
             if len(node.inputs) != len(candidate.inputs):
                 continue
 
+            cand_has_assert = False
+
+            # Get input list of the candidate with assert removed
+            cand_inputs_assert_removed = []
+            for i in candidate.inputs:
+                if i.owner and isinstance(i.owner.op,
+                                          theano.tensor.opt.Assert):
+                    cand_has_assert = True
+                    cand_inputs_assert_removed.append(i.owner.inputs[0])
+                else:
+                    cand_inputs_assert_removed.append(i)
+
+            # if the current node has assert input, it should not be
+            # replaced with a candidate node which has no assert input
```

Comment:
In that case, why not replace the candidate with the new node that have the assert?